[4.0] Meta Description character count

brianteeman opened the pull request:
redo of #25262 and #27359

A request was made to have a character count when entering a meta description.  I have done this and applied it to the Site Meta Description in the global configuration as shown in the screenshot and all the meta description fields)

![image](https://user-images.githubusercontent.com/1296369/71545212-84df4980-2980-11ea-8091-e0e99934bf17.png)


This can be applied to any textarea field by adding `charcounter="true"` to the field definition in the xml.

The character count has a default maximum of 200 characters but this can be configured in the field xml by adding `maxlength="160"` 

Accessibility is obviously very important and this PR addresses this as well. The meta description text is announced when the user focuses on the input together with the string to indicate the characters used. Obviously a live count announced by the screenreader would be distracting so it waits 2000 milliseconds before updating the user with the screen reader user with the current count.

The string `86 characters remaining of 160 characters.` is a regular joomla language string except the numbers are replaced  by the script. In this example I chose
`"{remaining} characters remaining of {maxlength} characters."`. Available placeholders are {remaining}, {maxlength}, {length}

### Documentation
Needs updating to document the new charcounter option

brianteeman committed:
[4.0] Meta Description character count

redo of #25262

brianteeman committed:
lint auto fix

richard67 commented on layouts/joomla/form/field/textarea.php:
```diff
@@ -42,12 +45,23 @@
+	// Load the js file
+	Factory::getDocument()->getWebAssetManager()->enableAsset('short-and-sweet');
+	// Set the css class to be used as the trigger
+	$charcounter = 'charcount';
```
```suggestion
	$charcounter = ' charcount';
```
I think later below in line 64 we need a space between `$class` and `$charcounter` for the case that `$class` is not empty, but only if $charcounter is not boolean false, that's why the space should not be added in line 64 between `$class` and `$charcounter` but here.

richard67 replied:
Or does `$class` already have a space at the end? If so, then forget my suggestions.

richard67 commented on layouts/joomla/form/field/textarea.php:
```diff
@@ -42,12 +45,23 @@
 	$columns ?: '',
 	$rows ?: '',
-	!empty($class) ? 'class="form-control ' . $class . '"' : 'class="form-control"',
+	!empty($class) ? 'class="form-control ' . $class . $charcounter . '"' : 'class="form-control ' . $charcounter . '"',
```
```suggestion
	!empty($class) ? 'class="form-control ' . $class . $charcounter . '"' : 'class="form-control' . $charcounter . '"',
```
If my above suggestion for line 57 is accepted we don't need the space here anymore for the case that `$class` is empty.

richard67 replied:
Or does `$class` already have a space at the end? If so, then forget my suggestions.

brianteeman committed:
new assets structure

richard67 commented:
@brianteeman I just tested with your last change for the asset structure. I get following JS error when opening global config:

`TypeError: menuToggleIcon is null, file admin-menu.js, line 72, column 7`.

brianteeman commented:
yeah - something isn't right - but too tired now - will look again in the morning

richard67 commented:
I can continue testing tomorrow evening after work.

richard67 commented:
Btw. my suggested changes are still valid. Last comment that you can forget about them applies only if value of variable “$class” has already a space at the end, but I don’t think that’s the case. Comment clearly says that but for a quick reader it might be less clear. Maybe my changes will even fix the js problem because they will fix css classes of the html element?

brianteeman commented:
@richard67 I am stepping through the code to see what causes the break - currently I am getting an error without the js changes I made 
Undefined method enableAsset in class Joomla\CMS\WebAsset\WebAssetManager

So I am guessing that something went wrong with the conflict fix.

richard67 commented:
@brianteeman Without this PR I don't get any JS error, and with this PR I get it only when going to Global Configuration where the character count is used.

What if the short-and-sweet js does a DOM query for css class "charcount" and doesn't find it due to the missing space between class names which I mentioned in my suggested changes, and there is no error handling?

I strongly suggest to apply my changes for a test and then see what happens.

brianteeman committed:
switched to helper instead of web asset

brianteeman commented:
I have fixed the space - it was a valid issue but not related to the problem you saw

I have switched to using htmlhelper to load the script for now as I couldnt work out how to use the webasset stuff.

As this is for review as a concept I think thats ok for now

thanks for all your help

richard67 commented on layouts/joomla/form/field/textarea.php:
```diff
@@ -42,12 +46,23 @@
 	$columns ?: '',
 	$rows ?: '',
-	!empty($class) ? 'class="form-control ' . $class . '"' : 'class="form-control"',
+	!empty($class) ? 'class="form-control ' . $class . $charcounter . '"' : 'class="form-control ' . $charcounter . '"',
```
```suggestion
	!empty($class) ? 'class="form-control ' . $class . $charcounter . '"' : 'class="form-control' . $charcounter . '"',
```
I think there is still this space which can be removed now because otherwise double space between "form-control" and "charcount" in the html. But that's a minor thing of course.

brianteeman replied:
missed that one

brianteeman committed:
space

Quy commented on layouts/joomla/form/field/textarea.php:
```diff
@@ -42,12 +46,23 @@
+	$charcounter = ' charcount';
+	// Set the text
+	$counterlabel = 'data-counter-label="' . Text::_('COM_CONFIG_METADESC_COUNTER') . '"';
+}
```
```suggestion
}

```

Quy commented on layouts/joomla/form/field/textarea.php:
```diff
@@ -58,7 +73,8 @@
 	$autofocus ? 'autofocus' : '',
 	$spellcheck ? '' : 'spellcheck="false"',
-	$maxlength ? $maxlength: ''
+	$maxlength ? $maxlength: '',
```
```suggestion
	$maxlength ? $maxlength : '',
```

Quy commented on layouts/joomla/form/field/textarea.php:
```diff
@@ -58,7 +73,8 @@
 	$autofocus ? 'autofocus' : '',
 	$spellcheck ? '' : 'spellcheck="false"',
-	$maxlength ? $maxlength: ''
+	$maxlength ? $maxlength: '',
```
```suggestion
	$maxlength ?: '',
```

brianteeman replied:
i am really not a fan of that notation

Quy replied:
See lines 63-64. It is for consistency but either way is fine with me.

brianteeman committed:
Update layouts/joomla/form/field/textarea.php

Co-Authored-By: Quy <quy@fluxbb.org>

brianteeman committed:
Update layouts/joomla/form/field/textarea.php

Co-Authored-By: Quy <quy@fluxbb.org>

richard67 commented:
@brianteeman No, thank YOU for all the work you do for Joomla .. I would be an a..hole if I could help but would not do.

Is it ready for testing?

brianteeman commented:
Yes

richard67 commented:
Should we ping Fedik to check if there is an issue with the web asset manager?

richard67 commented:
Ah, and by the way: Could we display the counter in hex instead of decimal? Those template ranters say our backend is made by developers for developers, and I don't wanna disappoint them ;-) (joking)

brianteeman commented:
Can't do hex but I can do binary

richard67 commented:
Even better ;-)

richard67 commented:
I have tested this item :white_check_mark: successfully on 483db5b25d28ae48d85749ded58a348aaceccc3e<br /><br />Works like a charm and looks good.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27804">issues.joomla.org/tracker/joomla-cms/27804</a>.</sub>

Quy commented on build/build-modules-js/settings.json:
```diff
@@ -492,6 +492,13 @@
           "dist/metismenujs.min.js": "js/metismenujs.min.js",
           "dist/metismenujs.min.js.map": "js/metismenujs.min.js.map"
+        }
+			},
```
```suggestion
      },
```

Quy replied:
Fix indentation.

jwaisner commented:
I have tested this item :white_check_mark: successfully on 483db5b25d28ae48d85749ded58a348aaceccc3e<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27804">issues.joomla.org/tracker/joomla-cms/27804</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27804">issues.joomla.org/tracker/joomla-cms/27804</a>.</sub>

brianteeman commented:
From the original post

> If accepted it is quick to add it to all the meta description fields)

richard67 commented:
@Fedik When this PR has been merged, could you have a look on assets handling and if possible in a future PR replace the htmlhelper calls here by usage of the new methods with asset manager?

richard67 commented:
@brianteeman Good question where to add it. Only all the meta descriptions? Or do we have other fields, too, where it could be useful?

I gave RTC because I assumed you will do the other fields with another, future PR, but if you wanna do it with this one here it is also OK. In the latter case: Shall I remove RTC?

brianteeman commented:
the only place to put it is the meta description. I will add them all to this PR tomorrow

richard67 commented:
@brianteeman So shall I remove RTC? Or leave it?

brianteeman committed:
add all components

brianteeman commented:
Now includes all components and updated original post

Quy commented:
Edit an article. 
Go to Publising tab.  
Below the textarea box.
String not translated. 
COM_CONFIG_METADESC_COUNTER

brianteeman commented:
Oops. Will correct in the morning by changing to a global string

brianteeman committed:
use a global string

brianteeman commented:
@quy language issues should be resolved now

Fedik committed:
Use WebAsset for counter (#98)

* use WebAsset

* fix js builder, to not lose attributes

brianteeman commented:
Please retest as @Fedik has provided the correct js changes to use webassets

richard67 commented:
Will retest tonight after work (CET). Thanks @Fedik .

jwaisner commented:
After applying this PR the article layout is messed up and I am showing errors.

![27804](https://user-images.githubusercontent.com/12568201/73953959-c25fbc80-48c6-11ea-8716-02ae4fb0afdf.png)

I ran npm ci as well. I also applied this PR via Patchtester and the diff.

Not sure if this a problem, but I noticed that the `package.json` and `package-lock.json` files were modified when applying the PR.

richard67 commented:
> Not sure if this a problem, but I noticed that the `package.json` and `package-lock.json` files were modified when applying the PR.

@jwaisner Is not a problem, is desired. This PR modifies them because introducing a new 3rd party npm package named "short-and-sweet" for the character counting.

brianteeman committed:
Update build/build-modules-js/settings.json

Co-Authored-By: Quy <quy@fluxbb.org>

Quy commented:
I have tested this item :white_check_mark: successfully on a73e030ce9294157a6cea095a45205ac03ad59f3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27804">issues.joomla.org/tracker/joomla-cms/27804</a>.</sub>

jwaisner commented:
I still have the same issue after the latest commit... I am backing out of testing this one as it is apparent I am unable to understand why this PR behaves any differently than others I have tested.

richard67 commented:
@jwaisner The message said that there is no script for the short-and-sweet package. Which version of patchtester did you use? Latest 4 beta still has issues when having a PR like this here which changes package-lock.json. Or 3? If 3: Have you run `npm install`? It will see the changed `package.json` and fetch the package and later build and copy the assets includinh the ones of this new package, and then it will update package-lock.json, too.

jwaisner commented:
@richard67 I am using 4.0 beta 2. Should I fall back to 3.0?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27804">issues.joomla.org/tracker/joomla-cms/27804</a>.</sub>

richard67 commented:
@jwaisner If you have npm: Yes, then apply patch and run "npm install". Then clear cache just to be sure.

richard67 commented:
I have tested this item :white_check_mark: successfully on a73e030ce9294157a6cea095a45205ac03ad59f3<br /><br />- Global Config - Site
- Category
- Contact
- Article
- Content Language
- Newsfeed
- Tags

All ok.

No errors in PHP log or browser console, no unwanted side effects noticed.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27804">issues.joomla.org/tracker/joomla-cms/27804</a>.</sub>

richard67 commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/27804">issues.joomla.org/tracker/joomla-cms/27804</a>.</sub>

wilsonge commented:
Thanks!

brianteeman commented:
Woohoo. Thanks all!!!

richard67 commented:
@brianteeman Shall I now make the PR to change to binary numbers?

kernusr commented:
@brianteeman Do you put a strict limit on the maximum length of the field? This will not save the form if the meta-description is longer than the set field length?
If so, then this approach is not true!
The limit on the maximum number of characters in the meta description, as in the browser header, is a recommendation and should not limit the user
